### PR TITLE
split test suite generation out of cbmc/jbmc

### PR DESCRIPTION
**Pull request by kroening**

### CODE REVIEW

File: src/Makefile [hunk 1]
```diff
@@ -1,10 +1,10 @@
-DIRS = ansi-c big-int cbmc jbmc cpp goto-cc goto-instrument goto-programs \
-       goto-symex langapi pointer-analysis solvers util linking xmllang \
-       assembler analyses java_bytecode \
+DIRS = ansi-c big-int cbmc ccover jbmc cpp goto-cc goto-instrument \
+       goto-programs goto-symex langapi pointer-analysis solvers util \
+       linking xmllang assembler analyses java_bytecode \
```
Comment (tautschnig) [inline src/Makefile]: To avoid future changes of this kind (reformatting of several lines just because one word changed), could each entry please be put on a separate line like other Makefiles now do?

File: src/ccover/ccover_bmc.cpp [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+
+\*******************************************************************/
+
+/// \file
+/// Bounded Model Checking for Test Suite Generation
+
+#include "ccover_bmc.h"
+
+#include <chrono>
+#include <iostream>
```
Comment (tautschnig) [inline src/ccover/ccover_bmc.cpp]: Is that actually required?

File: src/ccover/ccover_bmc.h [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+#ifndef CPROVER_CCOVER_CCOVER_BMC_H
+#define CPROVER_CCOVER_CCOVER_BMC_H
+
+#include <list>
+#include <map>
+
+#include <util/invariant.h>
+#include <util/options.h>
+#include <util/ui_message.h>
+#include <util/decision_procedure.h>
```
Comment (tautschnig) [inline src/ccover/ccover_bmc.h]: Required?

File: src/ccover/ccover_bmc.h [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+/// Bounded Model Checking for Test Suite Generation
+
+#ifndef CPROVER_CCOVER_CCOVER_BMC_H
+#define CPROVER_CCOVER_CCOVER_BMC_H
+
+#include <list>
+#include <map>
+
+#include <util/invariant.h>
+#include <util/options.h>
```
Comment (tautschnig) [inline src/ccover/ccover_bmc.h]: A forward declaration of `class optionst;` would suffice. `invariant.h` seems unused here.

File: src/ccover/ccover_bmc.h [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+#include <util/ui_message.h>
+#include <util/decision_procedure.h>
+
+#include <goto-programs/goto_trace.h>
+
+#include <goto-symex/symex_target_equation.h>
+#include <goto-programs/goto_model.h>
+#include <goto-programs/safety_checker.h>
+#include <goto-symex/memory_model.h>
```
Comment (tautschnig) [inline src/ccover/ccover_bmc.h]: I'm not sure all of the above are needed.

File: src/ccover/ccover_bmc.h [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+\*******************************************************************/
+
+/// \file
+/// Bounded Model Checking for Test Suite Generation
+
+#ifndef CPROVER_CCOVER_CCOVER_BMC_H
+#define CPROVER_CCOVER_CCOVER_BMC_H
+
+#include <list>
+#include <map>
```
Comment (tautschnig) [inline src/ccover/ccover_bmc.h]: I don't think either of those are used here, but `<memory>` is needed.

File: src/ccover/ccover_bmc.cpp [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+               << symex.total_vccs<<" VCC(s), "
+               << symex.remaining_vccs
+               << " remaining after simplification" << eom;
+}
+
+void ccover_bmct::setup_unwind()
+{
+  const std::string &set=options.get_option("unwindset");
+  std::vector<std::string> unwindset_loops;
+  split_string(set, ',', unwindset_loops, true, true);
```
Comment (tautschnig) [inline src/ccover/ccover_bmc.cpp]: Elsewhere a function called `parse_unwindset` exists and should be used, quite likely before this PR goes in.

File: src/ccover/ccover_bmc.h [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+  void get_memory_model();
+  void slice();
+
+  bool cover(
+    const goto_functionst &goto_functions,
+    const optionst::value_listt &criteria);
+
+  friend class bmc_covert;
+  template <template <class goalt> class covert>
+  friend class bmc_goal_covert;
```
Comment (tautschnig) [inline src/ccover/ccover_bmc.h]: Why are those friendships necessary?

File: src/ccover/ccover_parse_options.cpp [hunk 9]
```diff
@@ -0,0 +1,753 @@
...
+  {
+    v=unsafe_string2unsigned(cmdline.get_value("verbosity"));
+    if(v>10)
+      v=10;
+  }
+
+  ui_message_handler.set_verbosity(v);
+}
+
+void ccover_parse_optionst::get_command_line_options(optionst &options)
```
Comment (tautschnig) [inline src/ccover/ccover_parse_options.cpp:91]: CBMC has a much cleaned up command-line option parsing, could this please be made at least as good?

File: src/ccover/ccover_parse_options.h [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+
+/// \file
+/// CBMC Command Line Option Processing
+
+#ifndef CPROVER_CBMC_CBMC_PARSE_OPTIONS_H
+#define CPROVER_CBMC_CBMC_PARSE_OPTIONS_H
+
+#include <util/ui_message.h>
+#include <util/parse_options.h>
+//#include <util/timestamper.h>
```
Comment (tautschnig) [inline src/ccover/ccover_parse_options.h]: ?

File: src/ccover/ccover_parse_options.h [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+
+#include <langapi/language.h>
+
+//#include <analyses/goto_check.h>
```
Comment (tautschnig) [inline src/ccover/ccover_parse_options.h]: ?

File: src/cbmc/Makefile [hunk 12]
```diff
@@ -1,6 +1,5 @@
 SRC = all_properties.cpp \
       bmc.cpp \
-      bmc_cover.cpp \
```
Comment (tautschnig) [inline src/cbmc/Makefile:3]: This has already been moved in the previous commit, hence the prior commit yields an inconsistent state.

File: src/cbmc/cbmc_parse_options.cpp [hunk 13]
```diff
@@ -253,11 +249,14 @@ void cbmc_parse_optionst::get_command_line_options(optionst &options)
...
+  options.set_option(
+    "unwinding-assertions",
+    cmdline.isset("unwinding-assertions"));
 
-  if(cmdline.isset("partial-loops"))
-    options.set_option("partial-loops", true);
+  // generate unwinding assumptions otherwise
+  options.set_option(
+    "partial-loops",
+    cmdline.isset("partial-loops"));
```
Comment (tautschnig) [inline src/cbmc/cbmc_parse_options.cpp]: Seems unrelated and unnecessary?

File: src/jbmc/Makefile [hunk 14]
```diff
@@ -30,6 +27,9 @@ OBJ += ../ansi-c/ansi-c$(LIBEXT) \
       ../goto-instrument/cover_instrument_mcdc$(OBJEXT) \
       ../goto-instrument/cover_instrument_other$(OBJEXT) \
       ../goto-instrument/cover_util$(OBJEXT) \
+      ../goto-instrument/full_slicer$(OBJEXT) \
+      ../goto-instrument/reachability_slicer$(OBJEXT) \
+      ../goto-instrument/nondet_static$(OBJEXT) \
```
Comment (tautschnig) [inline src/jbmc/Makefile]: Ok, but unrelated?

**Review by tautschnig — CHANGES_REQUESTED** [review]

File: src/ccover/ccover_bmc.cpp [hunk 15]
```diff
@@ -0,0 +1,234 @@
...
+
+#include <goto-symex/memory_model_pso.h>
+#include <goto-symex/slice.h>
+#include <goto-symex/slice_by_trace.h>
+
+#include <cbmc/cbmc_solvers.h>
+
+#include <linking/static_lifetime_init.h>
+
+void ccover_bmct::do_conversion()
```
Comment (peterschrammel) [inline src/ccover/ccover_bmc.cpp:29]: Ideally, there should be a generic BMC class that provides these building block to avoid duplication.

File: src/ccover/ccover_parse_options.cpp [hunk 16]
```diff
@@ -0,0 +1,753 @@
...
+    "* *    University of Oxford Computer, Science Department    * *\n"
+    "* *                 kroening@kroening.com                   * *\n"
+    "\n"
+    "Usage:                       Purpose:\n"
+    "\n"
+    " ccover [-?] [-h] [--help]    show help\n"
+    " cccover file.c ...           source file names\n"
+    "\n"
+    "Test suite generation options:\n"
+    " --cover CC                   create test-suite with coverage criterion CC\n" // NOLINT(*)
```
Comment (peterschrammel) [inline src/ccover/ccover_parse_options.cpp:668]: What is the default behaviour if this option is not given?

File: src/ccover/ccover_bmc.cpp [hunk 17]
```diff
@@ -0,0 +1,234 @@
...
+    statistics() << "no slicing due to threads" << eom;
+  }
+  else
+  {
+    if(options.get_bool_option("slice-formula"))
+    {
+      ::slice(equation);
+      statistics() << "slicing removed "
+                   << equation.count_ignored_SSA_steps()
+                   << " assignments"<<eom;
```
Comment (peterschrammel) [inline src/ccover/ccover_bmc.cpp]: spaces missing around <<

File: src/ccover/ccover_parse_options.cpp (hunk 9 above)
Comment (peterschrammel) [reply src/ccover/ccover_parse_options.cpp:91]: Just as a note: Most of this code is duplicated in at least 8 different tools, which is not ideal. I'm not saying that needs to be fixed in this PR, but it has become difficult to distinguish intended from spurious differences between the command line interfaces.

**Comment by tautschnig** [comment]
I think this PR is right in setting a long-term direction, but as @peterschrammel also said: there should be de-duplication. In my opinion there should _first_ be one or more PRs to implement de-duplication, and then this one should follow.

File: src/Makefile (hunk 1 above)
Comment (kroening) [reply src/Makefile]: done

File: src/ccover/ccover_bmc.cpp (hunk 2 above)
Comment (kroening) [reply src/ccover/ccover_bmc.cpp]: done

File: src/ccover/ccover_bmc.h (hunk 3 above)
Comment (kroening) [reply src/ccover/ccover_bmc.h]: done

File: src/ccover/ccover_bmc.h (hunk 5 above)
Comment (kroening) [reply src/ccover/ccover_bmc.h]: done

File: src/ccover/ccover_bmc.h (hunk 6 above)
Comment (kroening) [reply src/ccover/ccover_bmc.h]: done

File: src/ccover/ccover_bmc.h (hunk 8 above)
Comment (kroening) [reply src/ccover/ccover_bmc.h]: done

File: src/jbmc/Makefile (hunk 14 above)
Comment (kroening) [reply src/jbmc/Makefile]: yes, simply sorting

File: src/cbmc/cbmc_parse_options.cpp (hunk 13 above)
Comment (kroening) [reply src/cbmc/cbmc_parse_options.cpp]: reverted

File: src/ccover/ccover_bmc.cpp (hunk 17 above)
Comment (kroening) [reply src/ccover/ccover_bmc.cpp]: done

File: src/ccover/ccover_parse_options.cpp (hunk 16 above)
Comment (kroening) [reply src/ccover/ccover_parse_options.cpp:668]: now documented

**Comment by smowton** [comment]
I note that all-properties (not-stop-on-fail) mode and cover mode have a *lot* in common, can they be factored?

File: src/ccover/ccover_bmc.cpp (hunk 15 above)
Comment (kroening) [reply src/ccover/ccover_bmc.cpp:29]: I don't see how this can be done in <2 years

**Comment by kroening** [comment]
I factored out the unwindset parsing.

**Comment by martin-cs** [comment]
I'm listed as a code owner but I think that's for about 4 lines of the change so I'm probably not the best person to review this.  Splitting out coverage and BMC may be an idea but this PR seems to cover a bunch of other things as well.

**Comment by NlightNFotis** [comment]
I am closing this PR as it seems to be very stale in terms of the current state of the branch it aims to merge to, and in terms of CI checks.

If you would like to see this revived, please feel free to reopen the PR, and continue working on this after it has been rebased on top of `develop`.
